### ISSUE TITLE
Need explanation about how the model works

### ISSUE THREAD

**Issue by gregory112** [kpot/keras-transformer#18]
I tried to recreate the model that you made in run_gpt.py with this simple code.

```python
x = np.array([[1, 1, 1], [1, 1, 1], [1, 1, 1], [1, 1, 1], [1, 1, 1]])
y = np.array([[1], [1], [1], [1], [1]])
y = np.expand_dims(y, axis=-1)
print(x.shape) # (5, 3)
print(y.shape) # (5, 1, 1)

input_layer = Input(shape=x.shape[1:], dtype='int32', name="input_layer")

transformer_layer = TransformerBlock(name="transformer", num_heads=2, use_masking=True, vanilla_wiring=False, residual_dropout=0.1, attention_dropout=0.1)

reusable_embedding_layer = ReusableEmbedding(10, 2, input_length=x.shape[1], name="reusable_embedding")

output_layer = TiedOutputEmbedding(projection_dropout=0.1, name='word_prediction_logits')

coordinate_embedding_layer = TransformerCoordinateEmbedding(1, name='coordinate_embedding')

transformer_act_layer = TransformerACT(name='adaptive_computation_time')

output_softmax_layer = Softmax(name='output_softmax')

next_step_input, embedding_matrix = reusable_embedding_layer(input_layer)
next_step_input = coordinate_embedding_layer(next_step_input, step=0)
next_step_input = transformer_layer(next_step_input)
next_step_input, act_output = transformer_act_layer(next_step_input)

transformer_act_layer.finalize()

final_output = output_softmax_layer(inputs=output_layer(inputs=[act_output, embedding_matrix]))

model = Model(inputs=[input_layer], outputs=[final_output])

optimizer = Adam(
        lr=0.1, beta_1=0.6, beta_2=0.999)

model.compile(
        optimizer,
        loss=sparse_categorical_crossentropy,
        metrics=["accuracy"])

model.summary(150)

model.fit(x, y) # This will return exception because x and y have different sizes

# print(model.predict([1, 1, 1]))
```

A simple universal transformer with ACT and depth=1. Looking at the model summary, the final output seems to be always tied to the number of elements in one sequence (in this case 3), and the number of possible classes (i.e vocabulary size), in this case 10. Why is it not outputting an output of 1x10? I mean, each position. I just don't get it. I don't understand the meaning of the output if interpreted like this. Output of the softmax are probability distribution, yes, I get it. But why is it 3x10 and not 1x10?

In case of an LSTM, the output is one per time step. I can just use the last output of the last timestep for example. So I thought with this model it should be the same. And there are other things that I think is a bit different with the paper (especially the paper described that the output of the encoder is with dimension of 512). This, that you implemented in the run_gpt.py I think is only the encoder?

Can you help me clearing things up?

**Comment by gregory112** [comment]
Never mind that, it is indeed the expected behavior of Transformer (especially the encoder block), which outputs a sequence representing the attended values of the input sequence. It is indeed sequence to sequence, not sequence to vector. My bad.

Some layers like mean, add, matmul, or even convolutaion can be added before softmax (or after) to "cut" the output sequence.
